common/src/types/v0/store/node.rs

```diff
@@ -66,33 +140,237 @@ impl NodeSpec {
     pub fn labels(&self) -> &NodeLabels {
         &self.labels
     }
+    /// Node labels
+    pub fn cordon_drain_state(&self) -> &Option<CordonDrainState> {
+        &self.cordon_drain_state
+    }
+
     /// Node gRPC endpoint
     pub fn set_endpoint(&mut self, endpoint: String) {
         self.endpoint = endpoint
     }
+
     /// Cordon node by applying the label.
     pub fn cordon(&mut self, label: String) {
-        self.cordon_labels.push(label);
+        // the the node has the label, return with an error
+        match &mut self.cordon_drain_state {
+            Some(ds) => match ds {
+                CordonDrainState::Cordoned(state) => {
+                    // just add the label
+                    state.add_cordon_label(label);
+                }
+                CordonDrainState::Draining(state) => {
+                    // just add the label
+                    state.add_cordon_label(label);
+                }
+                CordonDrainState::Drained(state) => {
+                    // just add the label
+                    state.add_cordon_label(label);
+                }
+            },
+            None => {
+                //add the label and set the state to cordoned
+                self.cordon_drain_state =
+                    Some(CordonDrainState::Cordoned(CordonedState::new(vec![label])));
+            }
+        }
+    }
+
+    /// Drain node by applying the drain label.
+    pub fn drain(&mut self, label: String) {
```

Comment:
nit: could drain be misinterpreted as draining the node `right now`? What about `set_drain`, or something like that?

Reply:
fixed